Custom highlighting themes

drmason13 opened the pull request:
Related to #419

Introduces once_cell dependency to store extra SyntaxSet, ThemeSet as statics.
Option<SyntaxSet> is no longer stored in the Config Markdown struct.

Gruvbox tmTheme added to test_site, it is taken from
https://github.com/Colorsublime/Colorsublime-Themes (MIT licensed)

It looks like I need to fix this up to merge into next (I wrote it against master originally, woops!).
I would appreciate a hand/pointer updating the documentation.

**IMPORTANT: Please do not create a Pull Request adding a new feature without discussing it first.**

The place to discuss new features is the forum: <https://zola.discourse.group/>
If you want to add a new feature, please open a thread there first in the feature requests section.

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -29,23 +35,79 @@ pub struct Markdown {
-    #[serde(skip_serializing, skip_deserializing)] // not a typo, 2 are need
-    pub extra_syntax_set: Option<SyntaxSet>,
+    /// A list of directories to search for additional `.tmTheme` files in.
+    pub extra_highlight_themes: Vec<String>,
```
I'm wondering if we should just add a single folder and search in that for additional syntaxes and themes instead of having 2 different vecs...

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -29,23 +35,79 @@ pub struct Markdown {
+        if let Some(theme) = &BUILTIN_HIGHLIGHT_THEME_SET.themes.get(&self.highlight_theme) {
+            theme
+        } else {
+            &EXTRA_HIGHLIGHT_THEME_SET.get().unwrap().themes[&self.highlight_theme]
```
We should probably invert that if/else in case someone wants to use a newer version of an existing theme for example.

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -29,23 +35,79 @@ pub struct Markdown {
+    }
+
+    /// Attempt to load any theme sets found in the extra highlighting themes of the config
+    /// TODO: move to markup.rs in 0.14
```
0.14 is the branch you did this PR on  and it is on markup.rs so you can remove that comment.

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -29,23 +35,79 @@ pub struct Markdown {
+    /// Attempt to load any theme sets found in the extra highlighting themes of the config
+    /// TODO: move to markup.rs in 0.14
+    pub fn load_extra_highlight_themes(&self, base_path: &Path) -> Result<Option<ThemeSet>> {
+        let extra_highlight_themes = self.extra_highlight_themes.clone();
```
I don't think you need to clone?

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -156,7 +148,12 @@ impl Config {
-        Config::parse(&content)
+
+        let config = Config::parse(&content)?;
+        let config_dir = path.parent().unwrap();
```
let's not unwrap if possible

Keats commented on components/config/src/highlighting.rs:
```diff
@@ -7,49 +8,60 @@ use syntect::parsing::SyntaxSet;
+                extra.find_syntax_by_token(lang)
+            })
+            .or_else(|| {
+                let hacked_lang = if lang == "js" || lang == "javascript" { "ts" } else { lang };
```
can you keep the comment?

Keats commented on docs/content/documentation/content/syntax-highlighting.md:
```diff
@@ -171,38 +171,85 @@ If your site source is laid out as follows:
+Tomorrow
+two-dark
+visual-studio-dark
+zenburn
```
there is already a list on the site: https://www.getzola.org/documentation/getting-started/configuration/#syntax-highlighting

Keats reviewed (changes requested):
What do you think of having a single folder for both additional syntaxes and themes rather than each having a vec?

meithecatte commented on components/config/src/highlighting.rs:
```diff
@@ -7,49 +8,60 @@ use syntect::parsing::SyntaxSet;
-pub enum HighlightSource {
-    Theme,
+pub static EXTRA_SYNTAX_SET: OnceCell<SyntaxSet> = OnceCell::new();
+pub static EXTRA_HIGHLIGHT_THEME_SET: OnceCell<ThemeSet> = OnceCell::new();
```
It is my understanding that you went for `OnceCell` because `ThemeSet` is not cloneable. I believe there is an easier way of resolving this problem: wrap it in an `Arc` instead.

drmason13 replied:
Hi, yes you are correct. I was working around clone. An Arc should work just as well, and without awkwardness around ensuring you only set it once.

drmason13 replied:
Ah, except `Arc<SyntaxSet>` doesn't impl Serialize/Deserialize. I don't know how to I would implement it manually, and I suspect it wouldn't be worth the hassle in this case?

drmason13 replied:
Done, sorted with skip_serialize/deserialize and `Arc<>`

drmason13 committed:
Support custom syntax highlighting themes

Related to #419

Gruvbox tmTheme added to test_site, it is taken from
https://github.com/Colorsublime/Colorsublime-Themes (MIT licensed)

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -43,26 +49,92 @@ pub struct Markdown {
+        if !THEME_SET.themes.contains_key(&self.highlight_theme) {
+            if let Some(extra) = &*self.extra_theme_set {
+                if !extra.themes.contains_key(&self.highlight_theme) {
+                    bail!(
```
I think it doesn't need to be a different error than the one below

Keats commented on components/rendering/benches/all.rs:
```diff
@@ -106,29 +106,30 @@ fn bench_render_content_without_highlighting(b: &mut test::Bencher) {
         &tera,
         &config,
-        "",
+        lang,
```
too many arguments eh :(

drmason13 replied:
It's a tad confusing yes!

Keats commented on components/config/src/config/markup.rs:
```diff
@@ -43,26 +49,92 @@ pub struct Markdown {
+        ))
+    }
+
+    pub fn export_theme_css(&self, theme_name: &str) -> String {
```
I would prefer if that stayed with the rest of the highlighting code so syntect doesn't bleed too much in the config but it doesn't seem doable anymore :(

Keats commented on components/site/src/lib.rs:
```diff
@@ -74,7 +73,7 @@ impl Site {
         let config_file = config_file.as_ref();
         let mut config = get_config(config_file)?;
-        config.markdown.load_extra_syntaxes(path)?;
+        config.markdown.load_extra_syntaxes_and_highlight_themes(path)?;
```
it should already be called by the line above no?

Keats reviewed (changes requested):
Just  a few things, I can merge and do it myself if you want
